File: tracks/r5-binary-parser/main.cpp
#include <array>
#include <cstdint>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <sstream>
#include <stdexcept>
#include <string>

struct Summary {
  uint32_t total_records = 0;
  uint64_t total_value = 0;
  std::array<uint32_t, 4> type_counts{};
  std::array<uint64_t, 4> type_value_sums{};
  uint32_t flagged_records = 0;
};

static uint16_t read_u16_le(const unsigned char* p) {
  return static_cast<uint16_t>(p[0]) |
         (static_cast<uint16_t>(p[1]) << 8);
}

static uint32_t read_u32_le(const unsigned char* p) {
  return static_cast<uint32_t>(p[0]) |
         (static_cast<uint32_t>(p[1]) << 8) |
         (static_cast<uint32_t>(p[2]) << 16) |
         (static_cast<uint32_t>(p[3]) << 24);
}

int main(int argc, char** argv) {
  if (argc != 2) {
    std::cerr << "usage: " << argv[0] << " <frames.bin>\n";
    return 1;
  }

  std::ifstream in(argv[1], std::ios::binary);
  if (!in) {
    std::cerr << "failed to open input\n";
    return 1;
  }

  Summary summary;
  std::array<unsigned char, 8> buf{};

  while (true) {
    in.read(reinterpret_cast<char*>(buf.data()), buf.size());
    std::streamsize n = in.gcount();
    if (n == 0) break;
    if (n != static_cast<std::streamsize>(buf.size())) {
      std::cerr << "truncated record\n";
      return 1;
    }

    uint16_t type = read_u16_le(buf.data());
    uint16_t flags = read_u16_le(buf.data() + 2);
    uint32_t value = read_u32_le(buf.data() + 4);

    if (type == 0 || type >= summary.type_counts.size()) {
      std::cerr << "unsupported record type: " << type << "\n";
      return 1;
    }

    summary.total_records += 1;
    summary.total_value += value;
    summary.type_counts[type] += 1;
    summary.type_value_sums[type] += value;
    if ((flags & 0x1u) != 0) {
      summary.flagged_records += 1;
    }
  }

  std::ostringstream out;
  out << "{";
  out << "\"total_records\":" << summary.total_records << ",";
  out << "\"total_value\":" << summary.total_value << ",";
  out << "\"flagged_records\":" << summary.flagged_records << ",";
  out << "\"type_counts\":{";
  bool first = true;
  for (size_t i = 1; i < summary.type_counts.size(); ++i) {
    if (!first) out << ",";
    first = false;
    out << "\"type_" << i << "\":" << summary.type_counts[i];
  }
  out << "},\"type_value_sums\":{";
  first = true;
  for (size_t i = 1; i < summary.type_value_sums.size(); ++i) {
    if (!first) out << ",";
    first = false;
    out << "\"type_" << i << "\":" << summary.type_value_sums[i];
  }
  out << "}}";

  std::cout << out.str() << "\n";
  return 0;
}

File: tracks/r5-binary-parser/main.rs
use std::collections::BTreeMap;
use std::env;
use std::fs;
use std::process;

fn fail(message: &str) -> ! {
    eprintln!("{message}");
    process::exit(1);
}

fn read_u16_le(bytes: &[u8]) -> u16 {
    u16::from_le_bytes([bytes[0], bytes[1]])
}

fn read_u32_le(bytes: &[u8]) -> u32 {
    u32::from_le_bytes([bytes[0], bytes[1], bytes[2], bytes[3]])
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        fail(&format!("usage: {} <frames.bin>", args[0]));
    }

    let bytes = fs::read(&args[1]).unwrap_or_else(|_| fail("failed to open input"));
    if bytes.len() % 8 != 0 {
        fail("truncated record");
    }

    let mut total_records: u32 = 0;
    let mut total_value: u64 = 0;
    let mut flagged_records: u32 = 0;
    let mut type_counts: BTreeMap<String, u32> = BTreeMap::new();
    let mut type_value_sums: BTreeMap<String, u64> = BTreeMap::new();

    for record in bytes.chunks_exact(8) {
        let record_type = read_u16_le(&record[0..2]);
        let flags = read_u16_le(&record[2..4]);
        let value = read_u32_le(&record[4..8]);

        if !(1..=3).contains(&record_type) {
            fail(&format!("unsupported record type: {record_type}"));
        }

        let key = format!("type_{record_type}");
        total_records += 1;
        total_value += value as u64;
        *type_counts.entry(key.clone()).or_insert(0) += 1;
        *type_value_sums.entry(key).or_insert(0) += value as u64;
        if (flags & 0x1) != 0 {
            flagged_records += 1;
        }
    }

    println!(
        "{{\"total_records\":{total_records},\"total_value\":{total_value},\"flagged_records\":{flagged_records},\"type_counts\":{},\"type_value_sums\":{}}}",
        serde_like_map_u32(&type_counts),
        serde_like_map_u64(&type_value_sums)
    );
}

fn serde_like_map_u32(map: &BTreeMap<String, u32>) -> String {
    map.iter()
        .map(|(k, v)| format!("\"{k}\":{v}"))
        .collect::<Vec<_>>()
        .join(",")
        .chars()
        .collect::<String>()
        .pipe(|inner| format!("{{{inner}}}"))
}

fn serde_like_map_u64(map: &BTreeMap<String, u64>) -> String {
    map.iter()
        .map(|(k, v)| format!("\"{k}\":{v}"))
        .collect::<Vec<_>>()
        .join(",")
        .chars()
        .collect::<String>()
        .pipe(|inner| format!("{{{inner}}}"))
}

trait Pipe: Sized {
    fn pipe<F, T>(self, f: F) -> T
    where
        F: FnOnce(Self) -> T,
    {
        f(self)
    }
}

impl<T> Pipe for T {}
